### DISCUSSION TITLE
My opinion about the new error handling mechanism

### DISCUSSION THREAD

**Discussion by QzCurious** [ecyrbe/zodios#209] [Ideas]
It's start from [`isDefinedError()`](https://github.com/ecyrbe/zodios/blob/ec792840f5ce99bf70b57fb6da57bd5fa3e52cb2/src/zodios-error.utils.ts#L15-L33). I think it's proper to parse the error while [resolving response](https://github.com/ecyrbe/zodios/blob/ec792840f5ce99bf70b57fb6da57bd5fa3e52cb2/src/zodios.ts#L279).

It could be like:
```js
try {
  const res = await response;
} catch (err) {
  // if thrown, it's the errors we not bother to handle at this level (like memory error)
  if (!axios.isAxiosError(err)) throw err;
  if (!err.response) throw err;

  // more on getEndpointErrors later
  const errorSchemas = getEndpointErrors(api, method, path, err.response.status);
  if (errorSchemas.some((schema) => schema.safeParse(err.response.data).success)) {
    throw new ExpectedError(err);
  } else {
    throw new UnknownError(err); // or just throw err
  }
}
```

(To simplify it, I'm not considering alias)
Then `isDefinedError` only checks if `error instanceof ExpectedError` and is produced by given `method` and `path`. But NOT parsing the `error`

```ts
function isDefinedError(err: unknown, method: string, path: string): boolean {
  if (!(err instanceof ExpectedError)) return false;
  return !!getEndpointErrors(api, err.config.method, err.config.url, err.response.status).length;
}
```

### Why `getEndpointErrors` (plural) instead of [`findEndpointError`](https://github.com/ecyrbe/zodios/blob/ec792840f5ce99bf70b57fb6da57bd5fa3e52cb2/src/utils.ts#L84) (single)

```ts
function getEndpointErrors(
  api: ZodiosEndpointDefinitions,
  method: string,
  path: string,
  err: AxiosError
): ZodiosEndpointError[];
```

1. So we can define multiple error with the same status code (and I really need this)
```ts
makeErrors([
  { status: 400, schema: z.object({ error: z.enum(["invalid_format"]) }) }
  {
    status: 400,
    schema: z.object({
      error: z.enum(["fail_to_lock_the_resource", "require_lock_timeout"])
    })
  }
])
```
2. _Determining which schema the thrown error is_ is a duty of zodios user, not zodios.

**Comment by ecyrbe** [comment]
Hello, 
when you say you need the same error status being declared multiple times for the same endpoint, what is the use case.
Why not just zod [or](https://github.com/colinhacks/zod#or) or [union](https://github.com/colinhacks/zod#unions) ?

**Reply by ecyrbe** [reply]
If it's to just to combine declared errors, here what i would do and evoid multiple declarations:
```ts
const invalidErrorSchema = z.object({ error: z.enum(["invalid_format"]) });
const lockErrorSchema = z.object({
      error: z.enum(["fail_to_lock_the_resource", "require_lock_timeout"])
    });

makeErrors([
  { status: 400, schema: z.union([invalidErrorSchema, lockErrorSchema]) }
])
```

**Reply by QzCurious** [reply]
I'll bring a example repo later. But let me describe why I don't think zod [or](https://github.com/colinhacks/zod#or) or [union](https://github.com/colinhacks/zod#unions) suitable. Althought two distinct error are having the same status code, it's doesn't mean they are related. It's counter-intuitive and inconvenience to define two distinct errors in a centralized way.

More on the implementation. I'm saying [isDefinedError()](https://github.com/ecyrbe/zodios/blob/ec792840f5ce99bf70b57fb6da57bd5fa3e52cb2/src/zodios-error.utils.ts#L15-L33) with [findEndpointError](https://github.com/ecyrbe/zodios/blob/ec792840f5ce99bf70b57fb6da57bd5fa3e52cb2/src/utils.ts#L84). Determaining schema of a error by only status code is too over. Some backend even choose to make almost all response status 200, no matter whether they were error or not. Think about [another Q&A](https://github.com/ecyrbe/zodios/discussions/203#discussioncomment-3967944) we have, you've provided a simple way to make a logical error group. I need it as well for all auth related error (ex: token expired, logged out by another login, user being locked). After applying the auth error group to an endpoint. The endpoint would not able to reuse status codes that used by auth error group.

Another personal reason is, I think Zodios is a perfect tool to do the connecting to virtually any web api. I need it be as UX-ish as possible.

**Reply by ecyrbe** [reply]
i'm willing to add support for multiple error matching since the generated error types already support it. 
So i'm ok to align runtime with it.
  
But i want to keep this separed from zodios instance though. So Error validation will stay on the typeguard.

So the support will still be on `isErrorFromPath` or `isErrorFromAlias`. This can be a transparent change.

**Reply by QzCurious** [reply]
Apart from bundle size, could I ask what else is to concern? A error response is also a response, so I tought it okey to validate it while [resolving response](https://github.com/ecyrbe/zodios/blob/ec792840f5ce99bf70b57fb6da57bd5fa3e52cb2/src/zodios.ts#L279). And `isErrorFromPath` can still be separated form zodios instance.

I might be too curious to know the decision. I'll respect to your decision.

**Reply by QzCurious** [reply]
Here [a example](https://codesandbox.io/s/zodios-multiple-error-matching-improperly-at-runtime-cfio4w?file=/src/api/api.ts) to show my usecase of defining multiple error with the same status code for a endpoint.

Here is a quick view to the main point:
![image](https://user-images.githubusercontent.com/38932402/198303763-3c84091b-5f90-4f5b-acc4-bd7a0a7e0321.png)

**Reply by ecyrbe** [reply]
The concerns are about:
- complexity of zodios request
- paying for validation that will most likely not be used (only some rare use cases require error validation)
- if you use the type guard, it means you are likely to want to handle the error and then not rethrowing it, then technically it's the same to have validation on the type guard since you'll run validation only once in both cases.

**Reply by QzCurious** [reply]
> paying for validation that will most likely not be used (only some rare use cases require error validation)

This one catches me. Yes, I also don't catch every error of my request. It's reasonable.

**Comment by ecyrbe** [comment]
@QzCurious fix is available on `v10.4.3` of `@zodios/core`

**Reply by QzCurious** [reply]
Amazing! Can't wait to try it tomorrow.
